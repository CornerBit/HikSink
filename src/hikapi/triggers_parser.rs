use minidom::Element;
use serde::{Deserialize, Serialize};

use super::EventIdentifier;

#[derive(Debug, PartialEq, Eq, Deserialize, Serialize, Clone)]
pub struct TriggerItem {
    pub identifier: EventIdentifier,
    pub hik_id: String,
    pub description: String,
}

impl TriggerItem {
    pub fn parse(s: &str) -> Result<Vec<TriggerItem>, TriggerParseError> {
        let root: Element = s.parse()?;
        let event_triggers = root
            .get_child("EventNotification", minidom::NSChoice::Any)
            .unwrap_or(&root)
            .get_child("EventTriggerList", minidom::NSChoice::Any)
            .unwrap_or(&root)
            .children();
        let mut parsed = vec![];

        for event_trigger in event_triggers {
            let hik_id = event_trigger
                .get_child("id", minidom::NSChoice::Any)
                .ok_or_else(|| TriggerParseError::FieldMissing("id".to_string()))?
                .text();
            let event_type = event_trigger
                .get_child("eventType", minidom::NSChoice::Any)
                .ok_or_else(|| TriggerParseError::FieldMissing("eventType".to_string()))?
                .text();
            let description = event_trigger
                .get_child("eventDescription", minidom::NSChoice::Any)
                .map(|e| e.text())
                .unwrap_or_else(String::new);
            let channel = event_trigger
                .get_child("videoInputChannelID", minidom::NSChoice::Any)
                .or_else(|| {
                    event_trigger.get_child("dynVideoInputChannelID", minidom::NSChoice::Any)
                })
                .or_else(|| event_trigger.get_child("inputIOPortID", minidom::NSChoice::Any))
                .or_else(|| event_trigger.get_child("dynInputIOPortID", minidom::NSChoice::Any))
                .map(|e| e.text());

            let event_type = event_type
                .parse()
                .map_err(|e| TriggerParseError::EventTypeInvalid(event_type, e))?;
            let identifier = EventIdentifier::new(channel, event_type);

            parsed.push(TriggerItem {
                hik_id,
                identifier,
                description,
            })
        }

        Ok(parsed)
    }
}

impl From<EventIdentifier> for TriggerItem {
    fn from(e: EventIdentifier) -> Self {
        TriggerItem {
            description: String::new(),
            hik_id: format!(
                "{}{}",
                e.event_type.to_string(),
                e.channel
                    .as_ref()
                    .map(|c| format!("-{}", c))
                    .unwrap_or_default()
            ),
            identifier: e,
        }
    }
}

quick_error! {
    #[derive(Debug)]
    pub enum TriggerParseError {
        XmlInvalid(error: minidom::Error) {
            from()
        }
        ChannelMissing {
            display("Event should contain either channelID or dynChannelID but neither were found")
        }
        FieldMissing(field: String) {
            display("Field was expected but missing: {}", field)
        }
        EventTypeInvalid(provided: String, error: String) {
            display("Event type `{}` was incorrectly formatted: {}", provided, error)
        }
    }
}

#[cfg(test)]
mod test {
    use super::TriggerItem;
    const TRIGGERS_CAM: &str = include_str!("../../samples/triggers_cam.xml");
    const TRIGGERS_NVR: &str = include_str!("../../samples/triggers_nvr.xml");
    const TRIGGERS_PTZ: &str = include_str!("../../samples/triggers_ptz.xml");

    #[test]
    fn test_parse_camera_samples() {
        let parsed = TriggerItem::parse(TRIGGERS_CAM).unwrap();
        insta::assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_parse_nvr_samples() {
        let parsed = TriggerItem::parse(TRIGGERS_NVR).unwrap();
        insta::assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_parse_ptz_samples() {
        let parsed = TriggerItem::parse(TRIGGERS_PTZ).unwrap();
        insta::assert_yaml_snapshot!(parsed);
    }
}
