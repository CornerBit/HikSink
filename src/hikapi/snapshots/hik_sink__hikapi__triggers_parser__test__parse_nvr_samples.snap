---
source: src/hikapi/triggers_parser.rs
expression: parsed

---
- identifier:
    channel: "1"
    event_type: Io
  hik_id: IO-1
  description: ""
- identifier:
    channel: "2"
    event_type: Io
  hik_id: IO-2
  description: ""
- identifier:
    channel: "3"
    event_type: Io
  hik_id: IO-3
  description: ""
- identifier:
    channel: "4"
    event_type: Io
  hik_id: IO-4
  description: ""
- identifier:
    channel: "101"
    event_type: Io
  hik_id: IO-101
  description: ""
- identifier:
    channel: "201"
    event_type: Io
  hik_id: IO-201
  description: ""
- identifier:
    channel: "401"
    event_type: Io
  hik_id: IO-401
  description: ""
- identifier:
    channel: "701"
    event_type: Io
  hik_id: IO-701
  description: ""
- identifier:
    channel: "702"
    event_type: Io
  hik_id: IO-702
  description: ""
- identifier:
    channel: "1001"
    event_type: Io
  hik_id: IO-1001
  description: ""
- identifier:
    channel: "1"
    event_type: Motion
  hik_id: VMD-1
  description: ""
- identifier:
    channel: "2"
    event_type: Motion
  hik_id: VMD-2
  description: ""
- identifier:
    channel: "3"
    event_type: Motion
  hik_id: VMD-3
  description: ""
- identifier:
    channel: "4"
    event_type: Motion
  hik_id: VMD-4
  description: ""
- identifier:
    channel: "5"
    event_type: Motion
  hik_id: VMD-5
  description: ""
- identifier:
    channel: "6"
    event_type: Motion
  hik_id: VMD-6
  description: ""
- identifier:
    channel: "7"
    event_type: Motion
  hik_id: VMD-7
  description: ""
- identifier:
    channel: "8"
    event_type: Motion
  hik_id: VMD-8
  description: ""
- identifier:
    channel: "9"
    event_type: Motion
  hik_id: VMD-9
  description: ""
- identifier:
    channel: "10"
    event_type: Motion
  hik_id: VMD-10
  description: ""
- identifier:
    channel: "15"
    event_type: Motion
  hik_id: VMD-15
  description: ""
- identifier:
    channel: "1"
    event_type: Tamper
  hik_id: tamper-1
  description: ""
- identifier:
    channel: "2"
    event_type: Tamper
  hik_id: tamper-2
  description: ""
- identifier:
    channel: "3"
    event_type: Tamper
  hik_id: tamper-3
  description: ""
- identifier:
    channel: "4"
    event_type: Tamper
  hik_id: tamper-4
  description: ""
- identifier:
    channel: "5"
    event_type: Tamper
  hik_id: tamper-5
  description: ""
- identifier:
    channel: "6"
    event_type: Tamper
  hik_id: tamper-6
  description: ""
- identifier:
    channel: "7"
    event_type: Tamper
  hik_id: tamper-7
  description: ""
- identifier:
    channel: "8"
    event_type: Tamper
  hik_id: tamper-8
  description: ""
- identifier:
    channel: "9"
    event_type: Tamper
  hik_id: tamper-9
  description: ""
- identifier:
    channel: "10"
    event_type: Tamper
  hik_id: tamper-10
  description: ""
- identifier:
    channel: "15"
    event_type: Tamper
  hik_id: tamper-15
  description: ""
- identifier:
    channel: "1"
    event_type: VideoLoss
  hik_id: videoloss-1
  description: ""
- identifier:
    channel: "2"
    event_type: VideoLoss
  hik_id: videoloss-2
  description: ""
- identifier:
    channel: "3"
    event_type: VideoLoss
  hik_id: videoloss-3
  description: ""
- identifier:
    channel: "4"
    event_type: VideoLoss
  hik_id: videoloss-4
  description: ""
- identifier:
    channel: "5"
    event_type: VideoLoss
  hik_id: videoloss-5
  description: ""
- identifier:
    channel: "6"
    event_type: VideoLoss
  hik_id: videoloss-6
  description: ""
- identifier:
    channel: "7"
    event_type: VideoLoss
  hik_id: videoloss-7
  description: ""
- identifier:
    channel: "8"
    event_type: VideoLoss
  hik_id: videoloss-8
  description: ""
- identifier:
    channel: "9"
    event_type: VideoLoss
  hik_id: videoloss-9
  description: ""
- identifier:
    channel: "10"
    event_type: VideoLoss
  hik_id: videoloss-10
  description: ""
- identifier:
    channel: "11"
    event_type: VideoLoss
  hik_id: videoloss-11
  description: ""
- identifier:
    channel: "12"
    event_type: VideoLoss
  hik_id: videoloss-12
  description: ""
- identifier:
    channel: "13"
    event_type: VideoLoss
  hik_id: videoloss-13
  description: ""
- identifier:
    channel: "14"
    event_type: VideoLoss
  hik_id: videoloss-14
  description: ""
- identifier:
    channel: "15"
    event_type: VideoLoss
  hik_id: videoloss-15
  description: ""
- identifier:
    channel: "16"
    event_type: VideoLoss
  hik_id: videoloss-16
  description: ""
- identifier:
    channel: "1"
    event_type: FieldDetection
  hik_id: fielddetection-1
  description: ""
- identifier:
    channel: "2"
    event_type: FieldDetection
  hik_id: fielddetection-2
  description: ""
- identifier:
    channel: "3"
    event_type: FieldDetection
  hik_id: fielddetection-3
  description: ""
- identifier:
    channel: "4"
    event_type: FieldDetection
  hik_id: fielddetection-4
  description: ""
- identifier:
    channel: "5"
    event_type: FieldDetection
  hik_id: fielddetection-5
  description: ""
- identifier:
    channel: "6"
    event_type: FieldDetection
  hik_id: fielddetection-6
  description: ""
- identifier:
    channel: "7"
    event_type: FieldDetection
  hik_id: fielddetection-7
  description: ""
- identifier:
    channel: "8"
    event_type: FieldDetection
  hik_id: fielddetection-8
  description: ""
- identifier:
    channel: "9"
    event_type: FieldDetection
  hik_id: fielddetection-9
  description: ""
- identifier:
    channel: "10"
    event_type: FieldDetection
  hik_id: fielddetection-10
  description: ""
- identifier:
    channel: "15"
    event_type: FieldDetection
  hik_id: fielddetection-15
  description: ""
- identifier:
    channel: "1"
    event_type: LineDetection
  hik_id: linedetection-1
  description: ""
- identifier:
    channel: "2"
    event_type: LineDetection
  hik_id: linedetection-2
  description: ""
- identifier:
    channel: "3"
    event_type: LineDetection
  hik_id: linedetection-3
  description: ""
- identifier:
    channel: "4"
    event_type: LineDetection
  hik_id: linedetection-4
  description: ""
- identifier:
    channel: "5"
    event_type: LineDetection
  hik_id: linedetection-5
  description: ""
- identifier:
    channel: "6"
    event_type: LineDetection
  hik_id: linedetection-6
  description: ""
- identifier:
    channel: "7"
    event_type: LineDetection
  hik_id: linedetection-7
  description: ""
- identifier:
    channel: "8"
    event_type: LineDetection
  hik_id: linedetection-8
  description: ""
- identifier:
    channel: "9"
    event_type: LineDetection
  hik_id: linedetection-9
  description: ""
- identifier:
    channel: "10"
    event_type: LineDetection
  hik_id: linedetection-10
  description: ""
- identifier:
    channel: "15"
    event_type: LineDetection
  hik_id: linedetection-15
  description: ""
- identifier:
    channel: "2"
    event_type: FaceDetection
  hik_id: facedetection-2
  description: ""
- identifier:
    channel: "3"
    event_type: FaceDetection
  hik_id: facedetection-3
  description: ""
- identifier:
    channel: "4"
    event_type: FaceDetection
  hik_id: facedetection-4
  description: ""
- identifier:
    channel: "5"
    event_type: FaceDetection
  hik_id: facedetection-5
  description: ""
- identifier:
    channel: "6"
    event_type: FaceDetection
  hik_id: facedetection-6
  description: ""
- identifier:
    channel: "7"
    event_type: FaceDetection
  hik_id: facedetection-7
  description: ""
- identifier:
    channel: "8"
    event_type: FaceDetection
  hik_id: facedetection-8
  description: ""
- identifier:
    channel: "9"
    event_type: FaceDetection
  hik_id: facedetection-9
  description: ""
- identifier:
    channel: "10"
    event_type: FaceDetection
  hik_id: facedetection-10
  description: ""
- identifier:
    channel: "15"
    event_type: FaceDetection
  hik_id: facedetection-15
  description: ""
- identifier:
    channel: "7"
    event_type: AudioException
  hik_id: audioexception-7
  description: ""
- identifier:
    channel: "1"
    event_type: RegionEntrance
  hik_id: regionEntrance-1
  description: ""
- identifier:
    channel: "7"
    event_type: RegionEntrance
  hik_id: regionEntrance-7
  description: ""
- identifier:
    channel: "1"
    event_type: RegionExiting
  hik_id: regionExiting-1
  description: ""
- identifier:
    channel: "7"
    event_type: RegionExiting
  hik_id: regionExiting-7
  description: ""
- identifier:
    channel: "2"
    event_type: UnattendedBaggage
  hik_id: unattendedBaggage-2
  description: ""
- identifier:
    channel: "3"
    event_type: UnattendedBaggage
  hik_id: unattendedBaggage-3
  description: ""
- identifier:
    channel: "4"
    event_type: UnattendedBaggage
  hik_id: unattendedBaggage-4
  description: ""
- identifier:
    channel: "5"
    event_type: UnattendedBaggage
  hik_id: unattendedBaggage-5
  description: ""
- identifier:
    channel: "6"
    event_type: UnattendedBaggage
  hik_id: unattendedBaggage-6
  description: ""
- identifier:
    channel: "7"
    event_type: UnattendedBaggage
  hik_id: unattendedBaggage-7
  description: ""
- identifier:
    channel: "8"
    event_type: UnattendedBaggage
  hik_id: unattendedBaggage-8
  description: ""
- identifier:
    channel: "9"
    event_type: UnattendedBaggage
  hik_id: unattendedBaggage-9
  description: ""
- identifier:
    channel: "10"
    event_type: UnattendedBaggage
  hik_id: unattendedBaggage-10
  description: ""
- identifier:
    channel: "15"
    event_type: UnattendedBaggage
  hik_id: unattendedBaggage-15
  description: ""
- identifier:
    channel: "2"
    event_type: AttendedBaggage
  hik_id: attendedBaggage-2
  description: ""
- identifier:
    channel: "3"
    event_type: AttendedBaggage
  hik_id: attendedBaggage-3
  description: ""
- identifier:
    channel: "4"
    event_type: AttendedBaggage
  hik_id: attendedBaggage-4
  description: ""
- identifier:
    channel: "5"
    event_type: AttendedBaggage
  hik_id: attendedBaggage-5
  description: ""
- identifier:
    channel: "6"
    event_type: AttendedBaggage
  hik_id: attendedBaggage-6
  description: ""
- identifier:
    channel: "7"
    event_type: AttendedBaggage
  hik_id: attendedBaggage-7
  description: ""
- identifier:
    channel: "8"
    event_type: AttendedBaggage
  hik_id: attendedBaggage-8
  description: ""
- identifier:
    channel: "9"
    event_type: AttendedBaggage
  hik_id: attendedBaggage-9
  description: ""
- identifier:
    channel: "10"
    event_type: AttendedBaggage
  hik_id: attendedBaggage-10
  description: ""
- identifier:
    channel: "15"
    event_type: AttendedBaggage
  hik_id: attendedBaggage-15
  description: ""
- identifier:
    channel: "1"
    event_type: SceneChangeDetection
  hik_id: scenechangedetection-1
  description: ""
- identifier:
    channel: "2"
    event_type: SceneChangeDetection
  hik_id: scenechangedetection-2
  description: ""
- identifier:
    channel: "3"
    event_type: SceneChangeDetection
  hik_id: scenechangedetection-3
  description: ""
- identifier:
    channel: "4"
    event_type: SceneChangeDetection
  hik_id: scenechangedetection-4
  description: ""
- identifier:
    channel: "5"
    event_type: SceneChangeDetection
  hik_id: scenechangedetection-5
  description: ""
- identifier:
    channel: "6"
    event_type: SceneChangeDetection
  hik_id: scenechangedetection-6
  description: ""
- identifier:
    channel: "8"
    event_type: SceneChangeDetection
  hik_id: scenechangedetection-8
  description: ""
- identifier:
    channel: "9"
    event_type: SceneChangeDetection
  hik_id: scenechangedetection-9
  description: ""
- identifier:
    channel: "10"
    event_type: SceneChangeDetection
  hik_id: scenechangedetection-10
  description: ""
- identifier:
    channel: "15"
    event_type: SceneChangeDetection
  hik_id: scenechangedetection-15
  description: ""
- identifier:
    channel: ~
    event_type: DiskFull
  hik_id: diskfull
  description: ""
- identifier:
    channel: ~
    event_type: DiskError
  hik_id: diskerror
  description: ""
- identifier:
    channel: ~
    event_type: NicBroken
  hik_id: nicbroken
  description: ""
- identifier:
    channel: ~
    event_type: IpConflict
  hik_id: ipconflict
  description: ""
- identifier:
    channel: ~
    event_type: IllegalAccess
  hik_id: illaccess
  description: ""
- identifier:
    channel: ~
    event_type: RecordingFailure
  hik_id: recordingfailure
  description: ""

